Update focus scoping

alicebartlett opened the pull request.

alicebartlett committed:
Update focus scoping

JakeChampion commented on src/scss/_mixins.scss:
```diff
@@ -80,11 +80,15 @@
+	// Standardise focus styles. This is scoped because IE allows the body tag to have focus which is... odd
+	a,
+	button,
+	[tabindex]{
```
Have we thought about applying this default focus style to input, select and textarea nodes?

alicebartlett replied:
like we do on line 92?

alicebartlett committed:
Fix linting error

JakeChampion commented on src/scss/_mixins.scss:
```diff
@@ -80,11 +80,15 @@
-	// Standardise outline styles
-	:focus {
-		outline: 2px solid $o-normalise-focus-color;
+	// Standardise focus styles. This is scoped because IE allows the body tag to have focus which is... odd
```
Another approach could be `body *:focus`, that would mean anything focused which is a descendant of the body node.

alicebartlett replied:
The thing I'm concerned about is IE giving focus to other odd things, this seemed like a safer way to work around this problem.
Are you concerned we might miss other focussable element types?

JakeChampion replied:
That is my concern, yes, other focusable elements which don't require a tabindex attribute are: input, select, textarea, object, area;

Another alternative would be to keep the root `:focus` styling and override it for the html, body and readonly nodes:
```
html:focus, body:focus, input[readonly]:focus {
    outline: none
}
```

I saw that approach in the default stylesheet bundled within chromium. https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/css/html.css?l=990

rowanmanning replied:
Yeah I agree with using something like the following, after properly giving it some thought:

```
html:focus, body:focus, input[readonly]:focus {
    outline: none
}
```

(apparently you can't blockquote a code block?)

It's better for some elements to have focus when we don't want them to than the other way around, which would present a11y issues. So yep I reckon go with the blacklist approach Jake suggested.

alicebartlett replied:
thanks team!

alicebartlett committed:
Fix focus styles for IE

IE8/9/10/11 all set :focus for body tags which is... not what we want.
This prevents that from happening.

JakeChampion commented on src/scss/_mixins.scss:
```diff
@@ -80,16 +80,25 @@
-	// Standardise outline styles
-	:focus {
+	// Standardise focus styles.
+	&:focus {
```
I think we need to remove the ampersand as this is a base-level rule and base-level rules can't use a parent referencing selector because there is no parent selector.

alicebartlett committed:
Fix typo
